### PR TITLE
feat(benchmark-tools): Update benchmark-tool to support custom measurement benchmarks

### PR DESCRIPTION
**Pull request by chentong7**
## Description

We've run into several scenarios in the past where we want to have benchmark tests which report their own measurements of something (other than the execution time and memory usage that benchmark-tool currently supports), and have those measurements go to Kusto so we can track them through time.

POC: https://github.com/microsoft/FluidFramework/pull/20772

## Sample
~~~
  Op Size
spec.js:54
    Insert Nodes
spec.js:54
      Many Transactions
spec.js:54
        ✔ 100 small nodes in 100 transactions @CustomBenchmark (134ms)
spec.js:83
        ✔ 100 medium nodes in 100 transactions @CustomBenchmark (89ms)
spec.js:83
        ✔ 100 large nodes in 100 transactions @CustomBenchmark (78ms)
spec.js:83
      Single Transaction
spec.js:54
        ✔ 100 small nodes in 1 transaction @CustomBenchmark (66ms)
spec.js:83
        ✔ 100 medium nodes in 1 transaction @CustomBenchmark (54ms)
spec.js:83
        ✔ 100 large nodes in 1 transaction @CustomBenchmark (53ms)
spec.js:83
    Remove Nodes
spec.js:54
      Many Transactions
spec.js:54
        ✔ 100 small nodes in 100 transactions (63ms)
spec.js:83
        ✔ 100 medium nodes in 100 transactions (62ms)
spec.js:83
        ✔ 100 large nodes in 100 transactions (62ms)
spec.js:83
      Single Transaction
spec.js:54
        ✔ 100 small nodes in 1 transactions containing 1 removal of 100 nodes
spec.js:76
        ✔ 100 medium nodes in 1 transactions containing 1 removal of 100 nodes
spec.js:76
~~~

## Unit test

~~~
Writing test results relative to package to nyc/junit-report.xml and nyc/junit-report.json
.mocharc.cjs:11

spec.js:54
  `benchmarkCustom` function
spec.js:54
    uses `before` and `after`
spec.js:54
      ✔ test @CustomBenchmark
spec.js:76
      ✔ run BenchmarkCustom
spec.js:76
  2 passing (8ms)
~~~

### CODE REVIEW

**Comment by Josmithr** [comment]
Note that this package is versioned and published independently from others in our repo. In order to publish the changes made here, you will want to update the `package.json`'s version, and add notes to the package's `CHANGELOG.md` file about the changes.

File: tools/benchmark/api-report/benchmark.alpha.api.md [hunk 1]
```diff
@@ -109,6 +112,16 @@ export interface CustomBenchmark extends BenchmarkTimingOptions {
 // @public (undocumented)
 export type CustomBenchmarkArguments = MochaExclusiveOptions & CustomBenchmark & BenchmarkDescription;
 
+// @public
+export interface CustomBenchmarkOptions extends MochaExclusiveOptions {
+    // (undocumented)
```
Comment (Josmithr) [inline tools/benchmark/api-report/benchmark.alpha.api.md:117]: Docs please 😁

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+// This file is a reporter used with node, so depending on node is fine.
+/* eslint-disable import/no-nodejs-modules */
```
Comment (Josmithr) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:7]: For what it's worth, I believe this entire tool is intended to be run in Node, so we could probably disable these rules for the entire package in its `.eslintrc` file.

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+ * Custom mocha reporter for benchmarks that provide their own measurements.
+ *
+ * @remarks
+ * It can be used by passing the JavaScript version of this file to mocha's --reporter option.
+ *
+ * This reporter takes output from mocha events and prints a user-friendly version of the results, in addition
+ * to writing them to a file.
+ * The path of the output file can be controlled with --reporterOptions reportDir=<path>.
+ *
+ * This reporter is coupled to customOutputRunner.ts, and depends on how it emits the actual benchmark data.
```
Comment (Josmithr) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:34]: Is there a specific class or function we can link to here instead of referencing the ts file? That will be less fragile to future code changes.

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+// eslint-disable-next-line import/no-internal-modules
+import type { CustomBenchmarkResults } from "./mocha/customOutputRunner";
+// eslint-disable-next-line import/no-internal-modules
+import { getSuiteName } from "./mocha/mochaReporterUtilities";
+
+/**
+ * Custom mocha reporter for benchmarks that provide their own measurements.
+ *
+ * @remarks
+ * It can be used by passing the JavaScript version of this file to mocha's --reporter option.
```
Comment (Josmithr) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:28]: Are there mocha docs we can link to here that describes how to do this?

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+ *
+ * This reporter takes output from mocha events and prints a user-friendly version of the results, in addition
+ * to writing them to a file.
+ * The path of the output file can be controlled with --reporterOptions reportDir=<path>.
```
Comment (Josmithr) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:32]: Nit (to highlight code syntax):
```suggestion
 * The path of the output file can be controlled with `--reporterOptions reportDir=<path>`.
```

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+import { Runner, Suite, Test } from "mocha";
+
+import { isChildProcess } from "./Configuration";
+import { pad, getName } from "./ReporterUtilities";
+// eslint-disable-next-line import/no-internal-modules
+import type { CustomBenchmarkResults } from "./mocha/customOutputRunner";
+// eslint-disable-next-line import/no-internal-modules
+import { getSuiteName } from "./mocha/mochaReporterUtilities";
+
+/**
```
Comment (Josmithr) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:24]: Thanks for these docs! They look great! I left a couple of Josh nitpicks, but these are really great!

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+ * This reporter takes output from mocha events and prints a user-friendly version of the results, in addition
+ * to writing them to a file.
+ * The path of the output file can be controlled with --reporterOptions reportDir=<path>.
+ *
+ * This reporter is coupled to customOutputRunner.ts, and depends on how it emits the actual benchmark data.
+ *
+ * See https://mochajs.org/api/tutorial-custom-reporter.html for more information about custom mocha reporters.
+ */
+class MochaCustomBenchmarkReporter {
+	private readonly inProgressSuites: Map<string, [string, CustomBenchmarkResults][]> = new Map();
```
Comment (Josmithr) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:39]: It would be good to document both of these properties, even though they're private. Will help future devs working in here.

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+					break;
+				}
+				outputFriendlyBenchmarks.push({
+					testName,
+					testData,
+				});
+			}
+		}
+
+		// Use the suite name as a filename, but first replace non-alphanumerics with underscores
```
Comment (Josmithr) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:173]: Nit: probably worth adding a note explaining _why_ we want to do the filename modification here.

File: pnpm-lock.yaml [hunk 9]
```diff
@@ -10000,7 +10000,7 @@ importers:
         version: link:../../test/test-drivers
       '@fluid-tools/benchmark':
         specifier: ^0.48.0
-        version: 0.48.0
+        version: link:../../../tools/benchmark
```
Comment (Josmithr) [inline pnpm-lock.yaml:10003]: You'll want to revert this before we merge.

File: tools/benchmark/src/mocha/customOutputRunner.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @public
+ */
+export interface CustomBenchmarkOptions extends MochaExclusiveOptions {
+	title: string;
+	decription: BenchmarkDescription;
+	run: (reporter: IMeasurementReporter) => void | Promise<unknown>;
+}
+/**
+ * This is a wrapper for Mocha's `it` function which runs the specified function `options.runBenchmark`
```
Comment (Josmithr) [inline tools/benchmark/src/mocha/customOutputRunner.ts:22]: Should this be
```suggestion
 * This is a wrapper for Mocha's `it` function which runs the specified via {@link CustomBenchmarkOptions.run}
```
?

**Comment by chentong7** [comment]
> Note that this package is versioned and published independently from others in our repo. In order to publish the changes made here, you will want to update the `package.json`'s version, and add notes to the package's `CHANGELOG.md` file about the changes.

Updated. Thx!

File: tools/benchmark/api-report/benchmark.alpha.api.md (hunk 1 above)
Comment (chentong7) [reply tools/benchmark/api-report/benchmark.alpha.api.md:117]: Added

File: tools/benchmark/src/MochaCustomOutputReporter.ts (hunk 2 above)
Comment (chentong7) [reply tools/benchmark/src/MochaCustomOutputReporter.ts]: Right right. Updated the `.eslintrc` rule

File: pnpm-lock.yaml (hunk 9 above)
Comment (chentong7) [reply pnpm-lock.yaml]: Removed the OpSize test here. It's used to test locally. I won't merge it with benchmark-tool custom option PR.

File: tools/benchmark/src/mocha/customOutputRunner.ts (hunk 10 above)
Comment (chentong7) [reply tools/benchmark/src/mocha/customOutputRunner.ts]: Yes. Updated

File: tools/benchmark/src/MochaCustomOutputReporter.ts (hunk 8 above)
Comment (chentong7) [reply tools/benchmark/src/MochaCustomOutputReporter.ts:176]: I actually don't know if not cleanup file name. What does it look like. Let me try remove it...

File: tools/benchmark/CHANGELOG.md [hunk 11]
```diff
@@ -5,3 +5,4 @@
 ### ⚠ BREAKING CHANGES
 
 Update `typescript` dependency from `4.x` to `5.x`.
+Update `version` from `0.50.0` to `0.51.1`.
```
Comment (CraigMacomber) [inline tools/benchmark/CHANGELOG.md:8]: This is inside of the `## 0.50.0` section of this file. If you are going to release a new version, it should get its own section in this file, and it should document what changes you are making in that version other than just the version change.

File: tools/benchmark/package.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@fluid-tools/benchmark",
-	"version": "0.50.0",
+	"version": "0.51.1",
```
Comment (CraigMacomber) [inline tools/benchmark/package.json:3]: Why 0.51.1? If this is a breaking API change, it should go to 51.0, otherwise 50.1.

File: tools/benchmark/src/mocha/customOutputRunner.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+import { Test } from "mocha";
+
+import type { BenchmarkDescription, MochaExclusiveOptions } from "../Configuration";
+import { timer } from "../timer";
+
+/**
+ * Options to configure a benchmark that reports custom measurements.
+ *
+ * @internal
```
Comment (CraigMacomber) [inline tools/benchmark/src/mocha/customOutputRunner.ts:14]: It seems like all your new stuff is marked `@internal`. Personally I don't think that makes a lot of sense. We document that "There are no API stability guarantees for packages in the @fluid-experimental and @fluid-internal scopes." (source: https://fluidframework.com/docs/build/releases-and-apitags/ )

Since this package is an independently versioned and published package, all consumers (including within this repo) consume it via a "^" and use its public API. If you add internal APIs, there are no places where we could safely use those unless we switched to pinning an exact version. Also since this is a dev tool, we can make breaking API changes and bump the version accordingly anyway without any major downsides, so there is no reason to mess with separate internal APIs.

File: tools/benchmark/src/mocha/customOutputRunner.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+export interface CustomBenchmarkOptions extends MochaExclusiveOptions {
+	/**
+	 * Title about benchmark option
+	 */
+	title: string;
+
+	/**
+	 * Set of options to describe a benchmark.
+	 */
+	decription: BenchmarkDescription;
```
Comment (CraigMacomber) [inline tools/benchmark/src/mocha/customOutputRunner.ts:25]: If you want to aggregate all the various configuration into a single interface like CustomBenchmarkOptions, I suggest you do it like how the existing APIs in this package do. Rather than making callers do {title: "x", description: { type: ... } ...}}, you should extend the BenchmarkDescription interface like BenchmarkOptions does.

File: tools/benchmark/.eslintrc.cjs [hunk 15]
```diff
@@ -18,7 +18,7 @@ module.exports = {
 	},
 	overrides: [
 		{
-			files: ["src/test/**"],
+			files: ["src/**"],
```
Comment (CraigMacomber) [inline tools/benchmark/.eslintrc.cjs:21]: This change allows you to take a dependency on node.js in none test code, which is likely to break anyone using this library outside of node.js. I did extra work to ensure this package can be used in environments like browser: I would prefer to not violate that, so I like having this enforcement. While we currently do not support our reporters in those environments, so they disable this lint, I'd like to keep it enforce for all the other code.

Maybe filter this to just ["src/*Reporter.ts", "src/test/**"], and add a comment explaining the reasoning?

**Comment by CraigMacomber** [comment]
CI is failing: you should fix the issues it found.

I see new APIs being added, but I don't see any changes to the tests. New APIs should get tests to both show how the new APIs are used (helpful for future developers working on this, customers and reviewers) as well as help ensure the APIs actually work correctly (which also helps reviewers).

Without any tests or docs added to the readme, or an explanation of the new features in the changelog, its a bit hard to find a good spot to start evaluating the design.

Given that, I think I know whats going on in this change, and I don't think its the way to get to where we want to go.

Currently the problem as I understand it is that:
- We do not offer an API for reporting arbitrary data as benchmark results
- We can't implement such an API since our reporter is not general enough to support that.

Last time we had this problem when trying to add memory tests, the approach taken was to duplicate a bunch of code, then customize it for that use case.

**What I would like to see:**

Rather than adding a bunch of new code which is mostly duplicating code we already have two copies of (for runtime benchmarks and memory benchmarks), I think we should instead work toward deduplicating the existing code.

Before the memory stuff was added, we had one reporter, factored into two parts:
- Reporter: logic for the reporter containing everything that's not test runner specific, making it easy to author reporters for specific test runners
- MochaReporter: Repoter+Mocha specific bits.

This made it easy for customers using other test runners to create the reporters they needed (ex: to work with jest).

When the memory stuff was added, it added a second reporter. This made running the tests harder (need to configure memory tests with a different reporter) and it didn't split out the test framework independent parts, so supporting other test runners is much less straight forward. Ideally a user of our library could just add a memory or custom test, and not need to author new reporters for their test framework, reconfigure their testing command line to include those reporters etc.

I view that as technical debt we should pay off before we continue work on this package.

What I think we should do in this change is modify our existing Reporter and MochaReporter to be sufficiently general that they can handle other kinds of data, like memory or custom. Infact they should only support arbitrary custom data, since that should be general enough to handle all cases. Then we can delete the memory reporter, and get down to a single reporter instead or going up to three different ones.

Once thats complete, then (in a separate PR) we could add the feature of this custom benchmark API. Some of new API should actually be used in the implementation of the time and memory benchmarks as they should all be able to be written as helpers using the new general purpose custom API.

Then we are left with a single reporter easy to use with multiple test framework and built in support for mocha, a library for making tests which report data to it, and a collection of helper libraries to measuring different things, including time and memory.

That leaves us with a much simpler to use library, with a much more modular implementation that I would expect to be far easier to understand, maintain, test and extend.

I guess that rant amounts to a proposed design doc for this work.

File: tools/benchmark/src/MochaCustomOutputReporter.ts [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+		const suiteNameEscaped: string = suiteName.replace(/[^\da-z]/gi, "_");
+		const outputContentString: string = JSON.stringify(
+			{ suiteName, tests: outputFriendlyBenchmarks },
+			undefined,
+			4,
+		);
+
+		// If changing this or the result file logic in general,
```
Comment (WayneFerrao) [inline tools/benchmark/src/MochaCustomOutputReporter.ts:184]: I saw this comment above as well, should it be abstracted to a NOTE in the function?

File: tools/benchmark/src/mocha/customOutputRunner.ts (hunk 13 above)
Comment (alexvy86) [reply tools/benchmark/src/mocha/customOutputRunner.ts:14]: Small nit: benchmark is in scope `@fluid-tools`, not `@fluid-experimental` or `@fluid-internal`. I don't know if we have explicit documentation about what we do/don't guarantee for that scope.

I agree with `@internal` being unnecessary though.

File: tools/benchmark/src/MochaCustomOutputReporter.ts (hunk 16 above)
Comment (chentong7) [reply tools/benchmark/src/MochaCustomOutputReporter.ts:184]: This file is no longer needed. Remove it now.

File: tools/benchmark/.eslintrc.cjs (hunk 15 above)
Comment (chentong7) [reply tools/benchmark/.eslintrc.cjs:21]: Revert

File: tools/benchmark/CHANGELOG.md [hunk 17]
```diff
@@ -1,5 +1,9 @@
 # @fluid-tools/benchmark
 
+## 0.52.0
+
+Provide @benchmarkCustom feature
```
[the remark this answers is not in the document]
Comment (chentong7) [reply tools/benchmark/CHANGELOG.md:5]: Gotcha. Add this change as part of version 0.51.0

File: tools/benchmark/src/mocha/customOutputRunner.ts [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+import type { BenchmarkDescription, MochaExclusiveOptions } from "../Configuration";
+import { timer } from "../timer";
+
+/**
+ * Options to configure a benchmark that reports custom measurements.
+ *
+ * @public
+ */
+export interface CustomBenchmarkOptions extends BenchmarkDescription, MochaExclusiveOptions {
```
[the remark this answers is not in the document]
Comment (chentong7) [reply tools/benchmark/src/mocha/customOutputRunner.ts:16]: Updated! Hmm..Should I also add `HookArguments` to extends so we support before() and after()?

[the remark this answers is not in the document]
Comment (alexvy86) [reply tools/benchmark/src/mocha/customOutputRunner.ts]: Yeah, that's actually a good point. I don't think we need them in the scenarios we have today but it's a good idea to make them available.

[the remark this answers is not in the document]
Comment (chentong7) [reply tools/benchmark/src/mocha/customOutputRunner.ts]: Awesome. Updated

File: tools/benchmark/src/mocha/customOutputRunner.ts [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+			addMeasurement: (key: string, value: number) => {
+				if (key in customData) {
+					throw new Error(`Measurement key '${key}' was already used.`);
+				}
+				customData[key] = value;
+			},
+		};
+
+		const startTime = timer.now();
+		await options.run(reporter);
```
Comment (alexvy86) [inline tools/benchmark/src/mocha/customOutputRunner.ts:62]: Since we're adding the hooks, we'll also need to call them. And that means a bit of change for `elapsedSeconds`, because we don't want to count the time spent in `after()`, so we can compute it here and just use the value below instead of computing it when creating the `results` object.
```suggestion
		await options.before?.();
		const startTime = timer.now();
		await options.run(reporter);
		const elapsedSeconds = timer.toSeconds(startTime, timer.now());
		await options.after?.();
```

File: tools/benchmark/CHANGELOG.md [hunk 20]
```diff
@@ -2,6 +2,8 @@
 
 ## 0.51.0
 
+Provide @benchmarkCustom feature
```
Comment (Josmithr) [inline tools/benchmark/CHANGELOG.md:5]: We should add more detail here. Detailing what `@benchmarkCustom` is and how to use it.

We should probably also update the package's README to include details about this new feature.

File: tools/benchmark/src/mocha/customOutputRunner.ts (hunk 19 above)
Comment (chentong7) [reply tools/benchmark/src/mocha/customOutputRunner.ts]: updated

File: tools/benchmark/CHANGELOG.md (hunk 20 above)
Comment (chentong7) [reply tools/benchmark/CHANGELOG.md]: updated

File: tools/benchmark/src/mocha/customOutputRunner.ts [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+	extends Titled,
+		BenchmarkDescription,
+		HookArguments,
+		MochaExclusiveOptions {
+	/**
+	 * Runs the benchmark.
+	 */
+	run: (reporter: IMeasurementReporter) => void | Promise<unknown>;
+}
+/**
```
Comment (CraigMacomber) [inline tools/benchmark/src/mocha/customOutputRunner.ts:32]: I generally like a blank line between top level items in a file.
```suggestion

/**
```

Comment (chentong7) [reply tools/benchmark/src/mocha/customOutputRunner.ts:33]: Me too. Update it

File: tools/benchmark/src/mocha/customOutputRunner.ts [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Options to configure a benchmark that reports custom measurements.
+ *
+ * @public
+ */
+export interface CustomBenchmarkOptions
+	extends Titled,
+		BenchmarkDescription,
+		HookArguments,
```
Comment (CraigMacomber) [inline tools/benchmark/src/mocha/customOutputRunner.ts:25]: On of the really awesome things about this new custom benchmark setup, is since the run function gets to control what is measured, and what is not, there is no need to support these extra hook arguments. Instead the caller can just do whatever is needed at the beginning and end of run.

This should make custom benchmarks much less annoying that our other benchmarks since passing data from setup into the measured part won't be a problem for users anymore.

```suggestion
```

**Review by CraigMacomber — COMMENTED** [review]
I left a suggested simplification.

Generally, though this is looking great. It's exciting to see reporter refactoring paying off. So clean :)

File: tools/benchmark/src/mocha/customOutputRunner.ts (hunk 22 above)
Comment (alexvy86) [reply tools/benchmark/src/mocha/customOutputRunner.ts:25]: Yeah I think Craig is right. I originally thought having before and after on the custom benchmarks made sense, but specifically for benchmark tests (as compared to regular mocha unit tests) one of the main uses for them seems to be just splitting the bits that we don't want to count towards the runtime measurement of a given benchmark... which is not the main point of the custom benchmarks. If those report a slightly higher runtime, I don't think that's necessarily bad or wrong. We'll still get the idea of how long it took to run the whole custom benchmark function.

The code bit that I suggested when we added HookArguments can be reverted too.

**Comment by chentong7** [comment]
> CI is failing: you should fix the issues it found.
> 
> I see new APIs being added, but I don't see any changes to the tests. New APIs should get tests to both show how the new APIs are used (helpful for future developers working on this, customers and reviewers) as well as help ensure the APIs actually work correctly (which also helps reviewers).
> 
> Without any tests or docs added to the readme, or an explanation of the new features in the changelog, its a bit hard to find a good spot to start evaluating the design.
> 
> Given that, I think I know whats going on in this change, and I don't think its the way to get to where we want to go.
> 
> Currently the problem as I understand it is that:
> 
> * We do not offer an API for reporting arbitrary data as benchmark results
> * We can't implement such an API since our reporter is not general enough to support that.
> 
> Last time we had this problem when trying to add memory tests, the approach taken was to duplicate a bunch of code, then customize it for that use case.
> 
> **What I would like to see:**
> 
> Rather than adding a bunch of new code which is mostly duplicating code we already have two copies of (for runtime benchmarks and memory benchmarks), I think we should instead work toward deduplicating the existing code.
> 
> Before the memory stuff was added, we had one reporter, factored into two parts:
> 
> * Reporter: logic for the reporter containing everything that's not test runner specific, making it easy to author reporters for specific test runners
> * MochaReporter: Repoter+Mocha specific bits.
> 
> This made it easy for customers using other test runners to create the reporters they needed (ex: to work with jest).
> 
> When the memory stuff was added, it added a second reporter. This made running the tests harder (need to configure memory tests with a different reporter) and it didn't split out the test framework independent parts, so supporting other test runners is much less straight forward. Ideally a user of our library could just add a memory or custom test, and not need to author new reporters for their test framework, reconfigure their testing command line to include those reporters etc.
> 
> I view that as technical debt we should pay off before we continue work on this package.
> 
> What I think we should do in this change is modify our existing Reporter and MochaReporter to be sufficiently general that they can handle other kinds of data, like memory or custom. Infact they should only support arbitrary custom data, since that should be general enough to handle all cases. Then we can delete the memory reporter, and get down to a single reporter instead or going up to three different ones.
> 
> Once thats complete, then (in a separate PR) we could add the feature of this custom benchmark API. Some of new API should actually be used in the implementation of the time and memory benchmarks as they should all be able to be written as helpers using the new general purpose custom API.
> 
> Then we are left with a single reporter easy to use with multiple test framework and built in support for mocha, a library for making tests which report data to it, and a collection of helper libraries to measuring different things, including time and memory.
> 
> That leaves us with a much simpler to use library, with a much more modular implementation that I would expect to be far easier to understand, maintain, test and extend.
> 
> I guess that rant amounts to a proposed design doc for this work.

Link the refactoring PR here: https://github.com/microsoft/FluidFramework/pull/21730

**Review by alexvy86 — APPROVED** [review]

File: tools/benchmark/README.md [hunk 23]
```diff
@@ -85,6 +86,12 @@ when you define the test) and `@ExecutionTime` (as opposed to `@MemoryUsage` for
 > The problem can be alleviated but not fully fixed using the `onCycle` hook argument.
 > See documentation on `HookArguments` for more detail.
 
+## Profiling custom usage
+
+To customize profiling, define tests using the benchmarkCustom() function. The run argument of this function
```
Comment (Josmithr) [inline tools/benchmark/README.md:91]: Nit
```suggestion
To customize profiling, define tests using the `benchmarkCustom()` function. The run argument of this function
```

**Review by Josmithr — APPROVED** [review]

File: tools/benchmark/src/mocha/customOutputRunner.ts (hunk 22 above)
Comment (chentong7) [reply tools/benchmark/src/mocha/customOutputRunner.ts]: Updated accordingly.

**Review by CraigMacomber — APPROVED** [review]
